### PR TITLE
[window] Optimization aggregation window by segment tree

### PR DESCRIPTION
**Pull request by shenh062326**
This optimization is based on https://www.vldb.org/pvldb/vol8/p1058-leis.pdf paper and is similar to https://duckdb.org/2021/10/13/windowing.html

1. Description
* When the window frame size is relatively large, such as window frame size >= 16, simple aggregation (each record aggregates its window data) will have a large number of repeated calculations.
* When the window is not incremental, for example the window frame start and frame end can be changed, or the aggregation function is difficult to remove records, incrementalAggregation cannot be used for optimization. Only simple aggregations can be used.

2. The optimization idea based on segment trees to reduce the repeated calculation of aggregate functions and reduce the time complexity of calculating a record from O(n) to O(log n) when the window frame size changes arbitrarily.

3. Optimize scenarios
* There are window aggregate functions.
* Window frame size >= 16 (meaning that the row size in one partition should to be at least 16), and as the window frame size becomes larger, the optimization becomes better.

4. The following are the benchmark results. The results show that as the window window frame size becomes larger, the optimization becomes more obvious. When the window frame size is 2048, the optimization is more than 10 times.

```
============================================================================
[...]/window/benchmark/WindowBenchmark.cpp     relative  time/iter   iters/s
============================================================================
rawRowFrame(count_frame_16_k_array)                       330.49ms      3.03
segmentOptRowFrame(count_frame_16_k_array)                313.07ms      3.19
rawRowFrame(count_frame_64_k_array)                       369.08ms      2.71
segmentOptRowFrame(count_frame_64_k_array)                292.25ms      3.42
rawRowFrame(count_frame_256_k_array)                      381.35ms      2.62
segmentOptRowFrame(count_frame_256_k_array)               301.02ms      3.32
rawRowFrame(count_frame_512_k_array)                      449.17ms      2.23
segmentOptRowFrame(count_frame_512_k_array)               353.11ms      2.83
rawRowFrame(count_frame_2048_k_array)                        1.45s   688.97m
segmentOptRowFrame(count_frame_2048_k_array)              298.44ms      3.35
rawRowFrame(sum_frame_16_k_array)                         309.85ms      3.23
segmentOptRowFrame(sum_frame_16_k_array)                  311.58ms      3.21
rawRowFrame(sum_frame_64_k_array)                         361.58ms      2.77
segmentOptRowFrame(sum_frame_64_k_array)                  316.90ms      3.16
rawRowFrame(sum_frame_256_k_array)                        602.75ms      1.66
segmentOptRowFrame(sum_frame_256_k_array)                 410.00ms      2.44
rawRowFrame(sum_frame_512_k_array)                           1.08s   926.15m
segmentOptRowFrame(sum_frame_512_k_array)                 429.23ms      2.33
rawRowFrame(sum_frame_2048_k_array)                          5.49s   181.99m
segmentOptRowFrame(sum_frame_2048_k_array)                414.17ms      2.41
rawRowFrame(avg_frame_16_k_array)                         398.75ms      2.51
segmentOptRowFrame(avg_frame_16_k_array)                  355.38ms      2.81
rawRowFrame(avg_frame_64_k_array)                         365.24ms      2.74
segmentOptRowFrame(avg_frame_64_k_array)                  350.43ms      2.85
rawRowFrame(avg_frame_256_k_array)                        580.88ms      1.72
segmentOptRowFrame(avg_frame_256_k_array)                 337.23ms      2.97
rawRowFrame(avg_frame_512_k_array)                        938.85ms      1.07
segmentOptRowFrame(avg_frame_512_k_array)                 373.71ms      2.68
rawRowFrame(avg_frame_2048_k_array)                          4.25s   235.56m
segmentOptRowFrame(avg_frame_2048_k_array)                381.15ms      2.62
rawRowFrame(min_frame_16_k_array)                         320.73ms      3.12
segmentOptRowFrame(min_frame_16_k_array)                  312.86ms      3.20
rawRowFrame(min_frame_64_k_array)                         358.26ms      2.79
segmentOptRowFrame(min_frame_64_k_array)                  308.58ms      3.24
rawRowFrame(min_frame_256_k_array)                        467.94ms      2.14
segmentOptRowFrame(min_frame_256_k_array)                 309.83ms      3.23
rawRowFrame(min_frame_512_k_array)                        665.46ms      1.50
segmentOptRowFrame(min_frame_512_k_array)                 322.60ms      3.10
rawRowFrame(min_frame_2048_k_array)                          2.31s   432.04m
segmentOptRowFrame(min_frame_2048_k_array)                358.51ms      2.79
rawRowFrame(max_frame_16_k_array)                         327.33ms      3.05
segmentOptRowFrame(max_frame_16_k_array)                  334.73ms      2.99
rawRowFrame(max_frame_64_k_array)                         342.32ms      2.92
segmentOptRowFrame(max_frame_64_k_array)                  402.26ms      2.49
rawRowFrame(max_frame_256_k_array)                        552.13ms      1.81
segmentOptRowFrame(max_frame_256_k_array)                 298.60ms      3.35
rawRowFrame(max_frame_512_k_array)                        590.44ms      1.69
segmentOptRowFrame(max_frame_512_k_array)                 302.45ms      3.31
rawRowFrame(max_frame_2048_k_array)                          2.20s   455.02m
segmentOptRowFrame(max_frame_2048_k_array)                385.36ms      2.59
rawRowFrame(stddev_frame_16_k_array)                      319.41ms      3.13
segmentOptRowFrame(stddev_frame_16_k_array)               434.15ms      2.30
rawRowFrame(stddev_frame_64_k_array)                      378.43ms      2.64
segmentOptRowFrame(stddev_frame_64_k_array)               552.84ms      1.81
rawRowFrame(stddev_frame_256_k_array)                        1.08s   927.33m
segmentOptRowFrame(stddev_frame_256_k_array)              679.89ms      1.47
rawRowFrame(stddev_frame_512_k_array)                        1.07s   931.61m
segmentOptRowFrame(stddev_frame_512_k_array)              787.58ms      1.27
rawRowFrame(stddev_frame_2048_k_array)                       3.77s   265.07m
segmentOptRowFrame(stddev_frame_2048_k_array)             849.46ms      1.18
```

### CODE REVIEW

**Comment by mbasmanova** [comment]
@shenh062326 Curious if this is based on https://www.vldb.org/pvldb/vol8/p1058-leis.pdf paper and similar to https://duckdb.org/2021/10/13/windowing.html

CI is red. Would you take a look?

**Comment by shenh062326** [comment]
> @shenh062326 Curious if this is based on https://www.vldb.org/pvldb/vol8/p1058-leis.pdf paper and similar to https://duckdb.org/2021/10/13/windowing.html

Yes, the optimization idea is base on it.

> CI is red. Would you take a look?

OK.

**Comment by shenh062326** [comment]
> @shenh062326 Curious if this is based on https://www.vldb.org/pvldb/vol8/p1058-leis.pdf paper and similar to https://duckdb.org/2021/10/13/windowing.html

Yes, the optimization idea is base on it.

> CI is red. Would you take a look?

OK.

**Comment by aditi-pandit** [comment]
@shenh062326 : Thanks for adding this optimization. Had a few request for you to help review the code.

- Please can you write a small design for SegmentTree describing the structure and the functions you have written.
- Can SegmentTree logic be implemented in a common library in Velox code ? That could help in other places as well. If that is the case then it would be great to write a class for it with its tests so that we can validate its correctness.
- The code here seems to have this config flag-driven. Generally Velox follows the philosophy of runtime optimizations so make a decision as you start evaluating. Could we trigger the SegmentTree based evaluation at runtime based on some criteria ?

**Comment by shenh062326** [comment]
> * Please can you write a small design doc for SegmentTree describing the structure and the functions you have written.

Add a design doc for SegmentTree: https://docs.google.com/document/d/126MIwwKJA_557_UUDy3QRqbO79ta0veTLZgKGObwRFo/edit

> * Can SegmentTree logic be implemented in a common library in Velox code ? That could help in other places as well. If that is the case then it would be great to write a class for it with its tests so that we can validate its correctness.

I'm not sure if segment tree will be used elsewhere, maybe it can be implemented in a common library in Velox in the future when this situation arises

> * The code here seems to have this config flag-driven. Generally Velox follows the philosophy of runtime optimizations so make a decision as you start evaluating. Could we trigger the SegmentTree based evaluation at runtime based on some criteria ?

Currently, for each window partition, choose segmentTree algorithm rather than simpleAggregation if: 
If the window frame start and frame end is constant value,  and frame size is larger than min_frame_size_use_segment_tree (default value 16), 
If the window frame is not constant, compute the average frame size for the input rows, if the average frame size is larger than min_frame_size_use_segment_tree (default value 16), 

If min_frame_size_use_segment_tree no need,  we can use a constant value instead.

File: velox/exec/Window.cpp [hunk 1]
```diff
@@ -185,6 +189,25 @@ void Window::createWindowFunctions() {
     windowFrames_.push_back(
         createWindowFrame(windowNode_, windowNodeFunction.frame, inputType));
   }
+
+  vector_size_t numFuncs = windowFunctions_.size();
```
[the remark this answers is not in the document]
Comment (shenh062326) [reply velox/exec/Window.cpp]: Done.

**Comment by shenh062326** [comment]
@aditi-pandit  Can you take a look at this PR.

**Comment by aditi-pandit** [comment]
> @aditi-pandit Can you take a look at this PR.

@shenh062326 : Would be great if you could add some unit tests for the segment tree structure by itself as well. That would help me validate your code overall.

I'll spend more time on this in the next few days.

**Comment by aditi-pandit** [comment]
Thanks @shenh062326 for working on this.

> > * Please can you write a small design doc for SegmentTree describing the structure and the functions you have written.
> 
> Add a design doc for SegmentTree: https://docs.google.com/document/d/126MIwwKJA_557_UUDy3QRqbO79ta0veTLZgKGObwRFo/edit

Please can you give me permissions to comment on this doc.

> 
> > * Can SegmentTree logic be implemented in a common library in Velox code ? That could help in other places as well. If that is the case then it would be great to write a class for it with its tests so that we can validate its correctness.
> 
> I'm not sure if segment tree will be used elsewhere, maybe it can be implemented in a common library in Velox in the future when this situation arises.

I still feel it will be better to separate SegmentTree into its own class with an API, so that we can test it independently from WindowFunctions. WindowFunctions might not be able to test all its corner cases in a way that ensures we got coverage.
wdyt ?
> 
> > * The code here seems to have this config flag-driven. Generally Velox follows the philosophy of runtime optimizations so make a decision as you start evaluating. Could we trigger the SegmentTree based evaluation at runtime based on some criteria ?
> 
> Currently, for each window partition, choose segmentTree algorithm rather than simpleAggregation if:
> 
> * If the window frame start and frame end is constant value,  and frame size is larger than min_frame_size_use_segment_tree (default value 16),

> * If the window frame is not constant, compute the average frame size for the input rows, if the average frame size is larger than min_frame_size_use_segment_tree (default value 16),
> 

This sounds good.

> If min_frame_size_use_segment_tree no need, we can use a constant value instead.
Did you get this number from the benchmark runs ? In any case, having a config for it is fine imo.

**Comment by shenh062326** [comment]
> Thanks @shenh062326 for working on this.
> 
> Please can you give me permissions to comment on this doc.

Done.

> 
> I still feel it will be better to separate SegmentTree into its own class with an API, so that we can test it independently from WindowFunctions. WindowFunctions might not be able to test all its corner cases in a way that ensures we got coverage. wdyt ?
> 

OK, I will separate SegmentTree.

> Did you get this number from the benchmark runs ? In any case, having a config for it is fine imo.

Yes, I get the number from  benchmark. I will keep one config.

**Comment by shenh062326** [comment]
1. Rebase master to fix conflict.
2. Separate SegmentTree into its own class.
3. Remove config min_frame_size_use_segment_tree and add kMinFrameSizeUseSegmentTree.

**Comment by aditi-pandit** [comment]
Will work on this Tues/Wed.

**Comment by aditi-pandit** [comment]
Thanks @shenh062326. I left some questions on your doc.

Your code looks good overall. Though it would be great to separate this code into 2 PRs for easier review : 

- Segment tree class and tests in the first PR.
- The rest of the changes in this PR in a 2nd one.

**Pull request closed without merging**